src/main/java/edu/berkeley/cs186/database/index/BPlusTree.java

```diff
@@ -420,20 +433,43 @@ private void typecheck(DataBox key) {
 
     // Iterator ////////////////////////////////////////////////////////////////
     private class BPlusTreeIterator implements Iterator<RecordId> {
-        // TODO(proj2): Add whatever fields and constructors you want here.
+        private LeafNode leafNode;
+        private Iterator<RecordId> rids;
+        private DataBox key;
+
+        public BPlusTreeIterator(LeafNode leafNode, Iterator<RecordId> rids) {
+            this.leafNode = leafNode;
+            this.rids = rids;
+            this.key = null;
+        }
+
+        public void setKey(DataBox key) {
+            this.key = key;
+        }
 
         @Override
         public boolean hasNext() {
-            // TODO(proj2): implement
-
-            return false;
+            // TODO(proj2): implement DONE
+            while (!rids.hasNext()) {
```

Comment:
```suggestion
            while (!rids.hasNext() && leafNode.getRightSibling().isPresent()) {
            currentLeaf = currentLeaf.getRightSibling().get();
            if (searchKey != null) {
                rids = leafNode.scanGreaterEqual(searchKey);
            } else {
                rids = leafNode.scanAll();
            }
```

요런식으로 분기문을 줄여보는 방향도 좋을 것 같아요! 장단점이 있는 것 같아요. 제 코드는 더 짧지만 짜왕 코드는 더 확실하게 조건 처리된다는 느낌이 강합니다.

Reply:
오오 감사합니다